### DISCUSSION TITLE
Ariakit with Remix Validated Form

### DISCUSSION THREAD

**Discussion by kwiat1990** [ariakit/ariakit#1614] [Questions] [answered]
I use Remix along with Remix Validated Form for all forms and as I switched from Mantine to Ariakit, I’m refactoring my custom input components, which are based on Mantine’s input components.  (See: https://github.com/airjp73/remix-validated-form/discussions/121#discussioncomment-3057353)

With Ariakit I get typing errors though. For some reason I can’t really define `state` or `value` inside `getInputProp` on Ariakit checkbox component. (I didn’t tried radio yet but I think it should also fail). On the other hand using Mantine I could set all native attributes as well as Mantine’s specific ones. 

Does anyone has already tried to use Remix Validated Form with Ariakit?

Codesandbox: https://codesandbox.io/s/gifted-elgamal-xvlf1o?file=/app/routes/index.tsx

**Comment by diegohaz** [comment, marked as answer]
https://github.com/ariakit/ariakit/discussions/1614#discussioncomment-3126336

**Reply by kwiat1990** [reply]
@diegohaz, I added a link to codesandbox to my initial message.

**Reply by diegohaz** [reply]
@kwiat1990 The typing errors come from the `getInputProps` function. It looks like an issue with the remix-validated-form library. I'm not familiar with it. I guess you'll find a better answer there.

**Reply by kwiat1990** [reply]
@diegohaz I link this question on theirs discussion board but it’s somehow odd that it had worked with both native and custom properties in Mantine components.

**Reply by diegohaz** [reply]
I'm not familiar with Mantine either. But if you assign the result of `getInputProps()` to a variable, without spreading it directly onto a component (which I believe should be a valid usage), you get the same error.

```jsx
const props = getInputProps({
  type: 'checkbox',
  checked: false,
})
```

There's no Ariakit involved there. That's why I'm not sure we can help.

**Reply by diegohaz** [reply]
Apparently, `remix-validated-form` types don't automatically infer the prop types of components when they're defined with generic types (which is the case of Ariakit):

```tsx
function WithoutGenerics(props: InputHTMLAttributes<HTMLInputElement>) {
  return null;
}

function WithGenerics<T extends InputHTMLAttributes<HTMLInputElement>>(props: T) {
  return null;
}

// works
<WithoutGenerics {...getInputProps({ type: "checkbox", checked: false })} />

// error
<WithGenerics {...getInputProps({ type: "checkbox", checked: false })} />
```

I'd say `getInputProps` types should be less restrictive. They should accept any props and return the same object shape, letting the component handle the prop types.

**Reply by kwiat1990** [reply]
I have found among issues on Remix Validated Form some, which also report problems with generics. A solution for this could be to explicit set the type for it: 
```tsx
import { Input, InputProps } from '@chakra-ui/react';

<Input {...getInputProps<InputProps>({ id: name, ...rest })} />
```

@diegohaz with Ariakit I'm not sure how I could pass default checkbox type to `getInputProps`. When I do this:
```tsx
 {...getInputProps<typeof Checkbox>(...)
```

I get following error: `Type 'Component<CheckboxOptions<"input">>' has no properties in common with type 'MinimalInputProps'`

**Reply by diegohaz** [reply]
You would have to do something like this:
```tsx
import { ComponentProps, InputHTMLAttributes } from 'react'

type CheckboxProps = Omit<
  ComponentProps<typeof Checkbox>,
  'checked' | 'defaultChecked'
> &
  InputHTMLAttributes<HTMLInputElement>

getInputProps<CheckboxProps>({
  type: 'checkbox',
  state: checkbox,
  checked: false,
})
```
